packages/core/src/stateUtils.ts

```diff
@@ -1078,9 +1064,25 @@ function getTransitionDomain<
     return transition.source;
   }
 
-  const lcca = findLCCA(targetStates.concat(transition.source));
+  const involvedStates = targetStates.concat(transition.source);
+  const [head] = involvedStates;
 
-  return lcca;
+  let current = getPathFromRootToNode(head);
+
+  for (let i = 1; i < involvedStates.length; i++) {
+    const path = getPathFromRootToNode(involvedStates[i]);
+    current = current.filter((sn) => path.includes(sn));
+  }
+
+  const domain = current[current.length - 1];
```

Comment:
We sort of had a bug here previously. We could return `undefined` from here, whereas the annotated return type is `StateNode<TContext, TEvent> | null`. Everything was working OK as `.parent === domainButActuallyUndefined` (and similar) was also a good stop for most of our internal while loops etc.